fyrox-impl/src/scene/tilemap/tileset.rs

```diff
@@ -90,70 +94,1505 @@ impl From<VisitError> for TileSetResourceError {
+    }
+    /// Find every transform set tile handle in this set. A transform set tile reference is a tile
+    /// with data that includes `transform_tile` with some value.
+    /// The given function will be called as `func(source_tile, transform_tile)` where
+    /// `source_tile` is the handle of the tile containing the data.
+    pub fn rebuild_transform_sets(&mut self) {
+        if let Some(t) = self.0.as_loaded_mut() {
+            t.rebuild_transform_sets()
+        }
+    }
+    /// Find a texture from some material page to serve as a preview for the tile set.
+    pub fn preview_texture(&self) -> Option<TextureResource> {
+        self.as_loaded_ref()
+            .map(|t| t.preview_texture())
+            .unwrap_or_default()
+    }
+    /// Get the page at the given position.
+    pub fn get_page(&self, position: Vector2<i32>) -> Option<&TileSetPage> {
+        self.as_loaded_ref()
+            .map(|t| t.get_page(position))
+            .unwrap_or_default()
+    }
+    /// Get the page at the given position.
+    pub fn get_page_mut(&mut self, position: Vector2<i32>) -> Option<&mut TileSetPage> {
+        self.0
+            .as_loaded_mut()
+            .map(|t| t.get_page_mut(position))
+            .unwrap_or_default()
+    }
+    /// Returns true if the given handle points to a tile definition.
+    pub fn is_valid_tile(&self, handle: TileDefinitionHandle) -> bool {
+        self.as_loaded_ref()
+            .map(|t| t.is_valid_tile(handle))
+            .unwrap_or_default()
+    }
+    /// The tile at the given page and tile coordinates.
+    pub fn get_abstract_tile(
+        &self,
+        page: Vector2<i32>,
+        tile: Vector2<i32>,
+    ) -> Option<AbstractTile> {
+        self.as_loaded_ref()
+            .map(|t| t.get_abstract_tile(page, tile))
+            .unwrap_or_default()
+    }
+    /// The render data for the tile at the given handle after applying the transform.
+    pub fn get_transformed_render_data(
+        &self,
+        trans: OrthoTransformation,
+        handle: TileDefinitionHandle,
+    ) -> Option<TileRenderData> {
+        self.as_loaded_ref()
+            .map(|t| t.get_transformed_render_data(trans, handle))
+            .unwrap_or_else(|| Some(TileRenderData::missing_data()))
+    }
+    /// Return the `TileRenderData` needed to render the tile at the given handle.
+    /// The handle is redirected if it refers to a reference to another tile.
+    /// If a reference is redirected and the resulting handle does not point to a tile definition,
+    /// then `TileRenderData::missing_tile()` is returned to that an error tile will be rendered.
+    /// If the given handle does not point to a reference and it does not point to a tile definition,
+    /// then None is returned since nothing should be rendered.
+    pub fn get_tile_render_data(
+        &self,
+        stage: TilePaletteStage,
+        handle: TileDefinitionHandle,
+    ) -> Option<TileRenderData> {
+        self.as_loaded_ref()
```

Comment:
I see what you mean. Obviously we shouldn't be doing that in the tight inner loop of `TileMap::collect_render_data`. I will rearrange things so that `as_loaded_ref` is only called once in `collect_render_data`.

Reply:
I have introduced the concept of `OptionTileSet` which is an `Option<&mut TileSet>` that is wrapped in a structure that provides `TileSet` methods and default behavior for the `None` case. This way `as_loaded_ref` only needs to be called once.